### ISSUE TITLE
Implement an equivalent of Instructor Python library's hook system

### ISSUE THREAD

**Issue by vkryukov** [thmsmlr/instructor_ex#77]
**Context:** [Instructor Python library](https://github.com/instructor-ai/instructor), which was the spiritual inspiration for instructor, implements a [hooks system](https://python.useinstructor.com/concepts/hooks/),

> a powerful mechanism for intercepting and handling events during the completion and parsing process in the Instructor library.

`instructor_ex` might want to do the same.

**Rationale:** In the current implementation, there is no way to capture the raw output of an LLM, which in addition to completion contains useful information, for example, finish reason, number of input and output tokens, the exact model version used, or some additional idiosyncratic information (`safetyRatings` in Gemini's case).

Implementing our own hook system would allow the end user to capture and process this information, and have more control over the completion - for example, do up to `n` retries but stop if the costs exceeds a certain budget.

**Comment by vkryukov** [comment]
For what it's worth (and in the spirit of `instuctor_ex` being a library to work with LLM), here is Claude's Sonnet 3.5 New advice on the implementation, after I fed it documentation for both libraries. If you agree with the first suggestion (using telemetry), I can draft an implementation.

1. Use Telemetry Instead of Custom Hooks

The most Elixir-idiomatic way to handle events/hooks would be to use :telemetry. It's a de-facto standard for instrumentation in the Elixir ecosystem, used by Phoenix, Ecto, and many other libraries. 

This approach has several advantages:

- It follows established Elixir patterns (:telemetry is widely used)
- It's more flexible than a custom hook system (multiple subscribers, standard tooling)
- It integrates well with existing monitoring solutions (e.g., Prometheus)
- It's non-intrusive (users can ignore it if they don't need it)
- It provides structured metadata about events
- It handles timing measurements properly
- It's well-documented and familiar to Elixir developers

```elixir
defmodule Instructor do
  # Define event prefixes
  @events [
    [:instructor, :completion, :start],     # Before LLM call
    [:instructor, :completion, :stop],      # After successful LLM call
    [:instructor, :completion, :exception], # When LLM call fails
    [:instructor, :parse, :start],         # Before parsing response
    [:instructor, :parse, :stop],          # After successful parse
    [:instructor, :parse, :exception],     # When parsing fails
    [:instructor, :validation, :start],    # Before validation
    [:instructor, :validation, :stop],     # After successful validation
    [:instructor, :validation, :exception] # When validation fails
  ]

  # Helper to emit events
  defp telemetry_span(event_name, meta, fun) do
    start_time = System.monotonic_time()
    
    :telemetry.execute(
      [:instructor, event_name, :start],
      %{system_time: System.system_time()},
      meta
    )

    try do
      result = fun.()
      end_time = System.monotonic_time()
      
      :telemetry.execute(
        [:instructor, event_name, :stop],
        %{
          duration: end_time - start_time,
          system_time: System.system_time()
        },
        Map.put(meta, :result, result)
      )
      
      result
    catch
      kind, reason ->
        end_time = System.monotonic_time()
        
        :telemetry.execute(
          [:instructor, event_name, :exception],
          %{
            duration: end_time - start_time,
            system_time: System.system_time()
          },
          Map.merge(meta, %{
            kind: kind,
            error: reason,
            stacktrace: __STACKTRACE__
          })
        )
        
        :erlang.raise(kind, reason, __STACKTRACE__)
    end
  end
  ```
  
  2. Provide Helper Modules for Common Use Cases
  
  While Telemetry provides the core functionality, we can make it more user-friendly by providing helper modules:
  
  ```elixir
  defmodule Instructor.Telemetry do
  @moduledoc """
  Helper functions for attaching common telemetry handlers
  """
  
  def attach_cost_logger(config \\ []) do
    :telemetry.attach_many(
      "instructor-cost-logger",
      [
        [:instructor, :completion, :stop]
      ],
      &handle_completion/4,
      config
    )
  end
  
  defp handle_completion([:instructor, :completion, :stop], measurements, metadata, _config) do
    # Extract token counts, calculate costs etc
    Logger.info("LLM call completed", 
      duration_ms: System.convert_time_unit(measurements.duration, :native, :millisecond),
      tokens: metadata.tokens,
      cost: calculate_cost(metadata)
    )
  end
end
```

3. Integration Points in Main Module

Update the existing code to emit telemetry events at key points:

```elixir
defp do_chat_completion(response_model, params, config) do
  telemetry_span(:completion, %{
    model: params[:model],
    response_model: response_model,
    config: config
  }, fn ->
    with {:ok, raw_response, params} <- do_adapter_chat_completion(params, config),
         {result, raw_response} <- telemetry_span(:parse, %{}, fn ->
           {cast_all(model, params), raw_response}
         end),
         {result, _raw_response} <- telemetry_span(:validation, %{}, fn ->
           {call_validate(response_model, changeset, validation_context), raw_response}
         end) do
      {:ok, changeset |> Ecto.Changeset.apply_changes()}
    else
      # ... error handling ...
    end
  end)
end
```

4. Documentation and Usage Examples

Provide clear documentation showing how to use the telemetry events:

```elixir
# In user's application
defmodule MyApp.LLMMetrics do
  require Logger

  def setup do
    # Track completion durations
    :telemetry.attach_many(
      "llm-metrics",
      [
        [:instructor, :completion, :stop],
        [:instructor, :completion, :exception]
      ],
      &handle_event/4,
      nil
    )
  end

  def handle_event([:instructor, :completion, :stop], measurements, metadata, _config) do
    # Report metrics to monitoring system
    metrics_reporter().record([
      {:llm_completion_duration, measurements.duration},
      {:llm_tokens, metadata.tokens}
    ])
  end
end
```

**Comment by vkryukov** [comment]
Another option, along the same telemetry lines, would be to use an existing plugins that build telemetry for Req, such as https://hexdocs.pm/req_telemetry/ReqTelemetry.html, and build our solution on top of that.

However, that would only offer a limited functionality, and won't be as feature rich as the direct telemetry implementation outlined above.

**Comment by vkryukov** [comment]
I've implemented some basic telemetry here: #83 , which addresses my immediate need. I will close this issue as a hook system might be overkill, at least for my use case.
